micro-rdk/src/common/data_collector.rs

```diff
@@ -0,0 +1,352 @@
+
+use crate::proto::app::data_sync::v1::SensorData;
+
+use super::{
+    config::{AttributeError, Kind},
+    movement_sensor::MovementSensor,
+    robot::ResourceType,
+    sensor::{get_sensor_readings_data, SensorError},
+};
+
+use thiserror::Error;
+
+/// A DataCollectorConfig instance is a representation of an element
+/// of the list of "capture_methods" in the "attributes" section of a
+/// component's configuration JSON object as stored in app. Each element
+/// of "capture_methods" is meant to produce an instance of `DataCollector`
+/// as defined below
+#[derive(Debug, Clone)]
+pub struct DataCollectorConfig {
+    pub method: CollectionMethod,
+    pub capture_frequency_hz: f32,
+}
+
+impl TryFrom<&Kind> for DataCollectorConfig {
+    type Error = AttributeError;
+    fn try_from(value: &Kind) -> Result<Self, Self::Error> {
+        if !value.contains_key("method")? {
+            return Err(AttributeError::KeyNotFound("method".to_string()));
+        }
+        if !value.contains_key("capture_frequency_hz")? {
+            return Err(AttributeError::KeyNotFound(
+                "capture_frequency_hz".to_string(),
+            ));
+        }
+        let method_str: String = value
+            .get("method")?
+            .ok_or(AttributeError::KeyNotFound("method".to_string()))?
+            .try_into()?;
+        let capture_frequency_hz = value
+            .get("capture_frequency_hz")?
+            .ok_or(AttributeError::KeyNotFound(
+                "capture_frequency_hz".to_string(),
+            ))?
+            .try_into()?;
+        // TODO: Collectors that take arguments (ex. Board Analogs)
+        // let parameters: &Kind = if let Ok(Some(params)) = value.get("additional_params") {
+        //     params
+        // } else {
+        //     &Kind::NullValue(0)
+        // };
+        let method = match method_str.as_str() {
+            "Readings" => CollectionMethod::Readings,
+            "AngularVelocity" => CollectionMethod::AngularVelocity,
+            "LinearAcceleration" => CollectionMethod::LinearAcceleration,
+            "LinearVelocity" => CollectionMethod::LinearVelocity,
+            // TODO: Power Sensor and Board collectors
+            // "Voltage" => CollectionMethod::Voltage,
+            // "Current" => CollectionMethod::Current,
+            // "Analogs" => {
+            //     let analog_reader_name: String = parameters.get("reader_name")?.ok_or(AttributeError::KeyNotFound("reader_name".to_string()))?.try_into()?;
+            //     CollectionMethod::Analogs(analog_reader_name)
+            // },
+            _ => {
+                return Err(AttributeError::ConversionImpossibleError);
+            }
+        };
+        Ok(DataCollectorConfig {
+            method,
+            capture_frequency_hz,
+        })
+    }
+}
+
+/// A CollectionMethod is an enum whose values are associated with
+/// a method on one or more component traits
+#[derive(Debug, Clone)]
+pub enum CollectionMethod {
```

Comment:
This design is very centralized / closed / static, in that if a new value returning method was added to one of these component types, we would need to make a compensating change here before it could be used with data capture. I'm not necessarily saying that's a bad design: it makes it a lot easier to see what is going on. The other way to do this would be to build into each component some ability to enumerate the method names / return Fn objects for each. That'd result in a more decentralized and open design. That said, nothing prevents us from moving to such a design later. So I think on balance I'm fine with starting with the closed design.

Reply:
Yeah I had considered that open design and I do agree that I like it better in theory, but I did not want to grapple with the lifetime considerations for function objects, so the design decision was born partially out of laziness

Reply:
Laziness - one of the principle virtues of the programmer! I'm fine with the closed approach, for now.